src/app/clusters/messages-server/messages-server.cpp

```diff
@@ -1,29 +1,199 @@
+    CHIP_ERROR ReadActiveMessageIds(app::AttributeValueEncoder & aEncoder, Delegate * delegate);
+    CHIP_ERROR ReadFeatureFlagAttribute(EndpointId endpoint, app::AttributeValueEncoder & aEncoder, Delegate * delegate);
+};
+
+MessagesAttrAccess gMessagesAttrAccess;
+
+CHIP_ERROR MessagesAttrAccess::Read(const app::ConcreteReadAttributePath & aPath, app::AttributeValueEncoder & aEncoder)
+{
+    EndpointId endpoint = aPath.mEndpointId;
+    Delegate * delegate = GetDelegate(endpoint);
+
+    switch (aPath.mAttributeId)
+    {
+    case app::Clusters::Messages::Attributes::Messages::Id:
+        if (isDelegateNull(delegate, endpoint))
+        {
+            return aEncoder.EncodeEmptyList();
+        }
+
+        return ReadMessages(aEncoder, delegate);
+    case app::Clusters::Messages::Attributes::ActiveMessageIDs::Id:
+        if (isDelegateNull(delegate, endpoint))
+        {
+            return aEncoder.EncodeEmptyList();
+        }
+
+        return ReadActiveMessageIds(aEncoder, delegate);
+    case app::Clusters::Messages::Attributes::FeatureMap::Id:
+        if (isDelegateNull(delegate, endpoint))
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        return ReadFeatureFlagAttribute(endpoint, aEncoder, delegate);
+    default:
+        break;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR MessagesAttrAccess::ReadFeatureFlagAttribute(EndpointId endpoint, app::AttributeValueEncoder & aEncoder,
+                                                        Delegate * delegate)
+{
+    uint32_t featureFlag = delegate->GetFeatureMap(endpoint);
+    return aEncoder.Encode(featureFlag);
+}
+
+CHIP_ERROR MessagesAttrAccess::ReadMessages(app::AttributeValueEncoder & aEncoder, Delegate * delegate)
+{
+    return delegate->HandleGetMessages(aEncoder);
+}
+
+CHIP_ERROR MessagesAttrAccess::ReadActiveMessageIds(app::AttributeValueEncoder & aEncoder, Delegate * delegate)
+{
+    return delegate->HandleGetActiveMessageIds(aEncoder);
+}
+
+} // anonymous namespace
+
 bool emberAfMessagesClusterPresentMessagesRequestCallback(
     chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
     const chip::app::Clusters::Messages::Commands::PresentMessagesRequest::DecodableType & commandData)
 {
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    EndpointId endpoint = commandPath.mEndpointId;
+    Status status       = Status::Success;
+
+    auto & messageId      = commandData.messageID;
+    auto & priority       = commandData.priority;
+    auto & messageControl = commandData.messageControl;
+    auto & startTime      = commandData.startTime;
+    auto & duration       = commandData.duration;
+    auto & messageText    = commandData.messageText;
+    auto & responses      = commandData.responses;
+
+    Delegate * delegate = GetDelegate(endpoint);
+    VerifyOrExit(isDelegateNull(delegate, endpoint) != true, err = CHIP_ERROR_INCORRECT_STATE);
```

Comment:
```suggestion
    VerifyOrExit(!isDelegateNull(delegate, endpoint), err = CHIP_ERROR_INCORRECT_STATE);
```

same below. Generally `foo != true` or `foo == false` feels odd.

Reply:
Created issue to track this for a subsequent PR:
https://github.com/project-chip/connectedhomeip/issues/32015